File: src/main/java/org/pandaframework/ecs/util/bits/Bits.java
package org.pandaframework.ecs.util.bits;

import java.util.BitSet;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

/**
 * @author Ranie Jade Ramiso
 */
public final class Bits {
    private static final Map<BitSet, ImmutableBitSet> CACHE = new HashMap<>();
    private static final BitSet BUILDER = new BitSet();

    public static BitSet compose(Consumer<BitSet> consumer) {
        BUILDER.clear();
        consumer.accept(BUILDER);

        if (CACHE.containsKey(BUILDER)) {
            return CACHE.get(BUILDER);
        }

        final ImmutableBitSet clone = new ImmutableBitSet(BUILDER);

        CACHE.put(clone, clone);

        return clone;
    }

    private Bits() {
    }


    private static final class ImmutableBitSet extends BitSet {
        private ImmutableBitSet(BitSet bitSet) {
            super(bitSet.cardinality());
            super.or(bitSet);
        }

        @Override
        public void flip(int bitIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void flip(int fromIndex, int toIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void set(int bitIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void set(int bitIndex, boolean value) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void set(int fromIndex, int toIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void set(int fromIndex, int toIndex, boolean value) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void clear(int bitIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void clear(int fromIndex, int toIndex) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void clear() {
            throw new UnsupportedOperationException();
        }

        @Override
        public void and(BitSet set) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void or(BitSet set) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void xor(BitSet set) {
            throw new UnsupportedOperationException();
        }

        @Override
        public void andNot(BitSet set) {
            throw new UnsupportedOperationException();
        }

        @Override
        public Object clone() {
            throw new UnsupportedOperationException();
        }
    }
}

File: src/main/kotlin/org/pandaframework/ecs/util/bits/Bits.kt
package org.pandaframework.ecs.util.bits

import java.util.*

/**
 * @author Ranie Jade Ramiso
 */
internal class Bits private constructor(private val internal: BitSet) {

    companion object {
        private val builder = BitSet()
        private val zero = Bits(BitSet())
        private var cache: MutableMap<BitSet, Bits> = mutableMapOf(zero.internal to zero)

        operator fun invoke (): Bits {
            return zero
        }

        private fun and(first: Bits, second: Bits): Bits {
            return compose {
                it.or(first.internal)
                it.and(second.internal)
            }
        }

        private fun or(first: Bits, second: Bits): Bits {
            return compose {
                it.or(first.internal)
                it.or(second.internal)
            }
        }

        private fun compose(composer: (BitSet) -> Unit): Bits {
            builder.clear()
            composer(builder)
            val clone = builder.clone() as BitSet
            return cache.getOrPut(clone, { Bits(clone) })
        }
    }

    fun and(other: Bits): Bits = and(this, other)

    fun or(other: Bits): Bits = or(this, other)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        } else if (other?.javaClass != javaClass) {
            return false
        }

        other as Bits

        if (internal != other.internal) {
            return false
        }

        return true
    }

    override fun hashCode(): Int {
        return internal.hashCode()
    }
}
